File: api/ml-predict.py
import json
import os
from http.server import BaseHTTPRequestHandler
import numpy as np
import joblib

BASE_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
MODEL_PATH = os.path.join(BASE_DIR, 'ml', 'model.pkl')

pkg = joblib.load(MODEL_PATH)
rf = pkg['model_rf']
scaler = pkg['scaler']
classes = pkg['classes']

KEY_PAIRS = [
    (0, 4), (0, 8), (0, 12), (0, 16), (0, 20),
    (4, 8), (4, 12), (4, 16), (4, 20),
    (8, 12), (12, 16), (16, 20),
    (3, 7), (7, 11), (11, 15), (15, 19),
    (2, 6), (6, 10), (10, 14), (14, 18),
]

def compute_distances(pts):
    dists = []
    for i, j in KEY_PAIRS:
        d = np.linalg.norm(pts[i] - pts[j])
        dists.append(d)
    return np.array(dists)

def compute_angles(pts):
    angles = []
    for i, j, k in [(0, 4, 8), (4, 8, 12), (8, 12, 16), (12, 16, 20),
                    (0, 5, 9), (5, 9, 13), (9, 13, 17),
                    (1, 5, 9), (2, 6, 10), (3, 7, 11)]:
        v1 = pts[i] - pts[j]
        v2 = pts[k] - pts[j]
        norm = np.linalg.norm(v1) * np.linalg.norm(v2)
        if norm == 0:
            angles.append(0)
        else:
            cos_a = np.clip(np.dot(v1, v2) / norm, -1, 1)
            angles.append(np.arccos(cos_a))
    return np.array(angles)

def extract_features(row_vals):
    pts = np.array(row_vals).reshape(21, 3)
    raw = row_vals
    dists = compute_distances(pts)
    angles = compute_angles(pts)
    wrist = pts[0]
    rel = (pts - wrist).flatten()
    return np.concatenate([raw, rel, dists, angles])

class handler(BaseHTTPRequestHandler):
    def do_POST(self):
        content_length = int(self.headers.get('Content-Length', 0))
        body = self.rfile.read(content_length)
        try:
            data = json.loads(body)
            landmarks = data.get('landmarks')
            if not isinstance(landmarks, list) or len(landmarks) == 0:
                raise ValueError('landmarks array is required')
            feats = extract_features(landmarks)
            feats_s = scaler.transform(feats.reshape(1, -1))
            probs = rf.predict_proba(feats_s)[0]
            top3_idx = np.argsort(probs)[-3:][::-1]
            results = [{'label': classes[i], 'confidence': float(probs[i])} for i in top3_idx]
            response = json.dumps({'ok': True, 'predictions': results})
            self.send_response(200)
            self.send_header('Content-Type', 'application/json')
            self.end_headers()
            self.wfile.write(response.encode())
        except Exception as e:
            response = json.dumps({'error': 'ML prediction failed', 'message': str(e)})
            self.send_response(500)
            self.send_header('Content-Type', 'application/json')
            self.end_headers()
            self.wfile.write(response.encode())

File: api/ml-predict.js
import fs from 'fs'
import path from 'path'
import { fileURLToPath } from 'url'

const __dirname = path.dirname(fileURLToPath(import.meta.url))
const MODEL_PATH = path.join(__dirname, '..', 'ml', 'model.json')

const model = JSON.parse(fs.readFileSync(MODEL_PATH, 'utf-8'))

const KEY_PAIRS = [
  [0, 4], [0, 8], [0, 12], [0, 16], [0, 20],
  [4, 8], [4, 12], [4, 16], [4, 20],
  [8, 12], [12, 16], [16, 20],
  [3, 7], [7, 11], [11, 15], [15, 19],
  [2, 6], [6, 10], [10, 14], [14, 18],
]

const ANGLE_TRIPLES = [
  [0, 4, 8], [4, 8, 12], [8, 12, 16], [12, 16, 20],
  [0, 5, 9], [5, 9, 13], [9, 13, 17],
  [1, 5, 9], [2, 6, 10], [3, 7, 11],
]

function dist(a, b) {
  return Math.sqrt((a[0] - b[0]) ** 2 + (a[1] - b[1]) ** 2 + (a[2] - b[2]) ** 2)
}

function computeDistances(pts) {
  return KEY_PAIRS.map(([i, j]) => dist(pts[i], pts[j]))
}

function computeAngles(pts) {
  return ANGLE_TRIPLES.map(([i, j, k]) => {
    const v1 = [pts[i][0] - pts[j][0], pts[i][1] - pts[j][1], pts[i][2] - pts[j][2]]
    const v2 = [pts[k][0] - pts[j][0], pts[k][1] - pts[j][1], pts[k][2] - pts[j][2]]
    const dot = v1[0] * v2[0] + v1[1] * v2[1] + v1[2] * v2[2]
    const n1 = Math.sqrt(v1[0] ** 2 + v1[1] ** 2 + v1[2] ** 2)
    const n2 = Math.sqrt(v2[0] ** 2 + v2[1] ** 2 + v2[2] ** 2)
    if (n1 === 0 || n2 === 0) return 0
    const cos = Math.max(-1, Math.min(1, dot / (n1 * n2)))
    return Math.acos(cos)
  })
}

function extractFeatures(rowVals) {
  const pts = []
  for (let i = 0; i < 21; i++) {
    pts.push([rowVals[i * 3], rowVals[i * 3 + 1], rowVals[i * 3 + 2]])
  }
  const raw = rowVals
  const wrist = pts[0]
  const rel = pts.flatMap(p => [p[0] - wrist[0], p[1] - wrist[1], p[2] - wrist[2]])
  const dists = computeDistances(pts)
  const angles = computeAngles(pts)
  return [...raw, ...rel, ...dists, ...angles]
}

function predict(landmarks) {
  const feats = extractFeatures(landmarks)
  const scaled = feats.map((v, i) => (v - model.scaler_mean[i]) / model.scaler_scale[i])
  const probs = new Array(model.n_classes).fill(0)

  for (const tree of model.trees) {
    let node = 0
    while (tree.children_left[node] !== -1) {
      node = scaled[tree.feature[node]] <= tree.threshold[node]
        ? tree.children_left[node]
        : tree.children_right[node]
    }
    const counts = tree.value[node]
    for (let c = 0; c < counts.length; c++) {
      probs[c] += counts[c]
    }
  }

  const total = probs.reduce((s, v) => s + v, 0)
  const normalized = probs.map(v => v / total)
  const indices = normalized.map((v, i) => i).sort((a, b) => normalized[b] - normalized[a])

  return indices.slice(0, 3).map(i => ({
    label: model.classes[i],
    confidence: normalized[i],
  }))
}

export default function handler(req, res) {
  if (req.method !== 'POST') {
    return res.status(405).json({ error: 'Method not allowed' })
  }
  try {
    const { landmarks } = req.body || {}
    if (!Array.isArray(landmarks) || landmarks.length === 0) {
      return res.status(400).json({ error: 'landmarks array is required' })
    }
    if (landmarks.length !== 63) {
      return res.status(400).json({ error: 'landmarks must contain 63 values (21 × 3)' })
    }
    const predictions = predict(landmarks)
    res.json({ ok: true, predictions })
  } catch (error) {
    console.error('ML prediction failed:', error)
    res.status(500).json({ error: 'ML prediction failed', message: error.message })
  }
}
